### PR TITLE
(Scripts/AI) ZG20: Jeklik (Bat Boss) - Implement Bat Riders, other fixes and improvements

### PR DESCRIPTION
**Pull request by kjack9**
## Changes Proposed:

This PR proposes changes to:
-  [X] Core (units, players, creatures, game systems).
-  [X] Scripts (bosses, spell scripts, creature scripts).
-  [X] Database (SAI, creatures, etc).

This is a extensive updating of the Jeklik fight in ZG with particular focus on the bomb-throwing bats.

In Blizzard's infinite wisdom, they used the same creature ID (Gurubashi Bat Rider, 14750) for both the roaming trash packs and Jeklik's flying bats that drop firebombs. The previous iteration used a totally different creature, Frentzied Bloodseeker Bat (14965), to drop the bombs from a fixed position above the boss' head. As far as I can tell these creatures are unrelated to the real encounter on Classic.

This PR restores the Gurubashi Bat Riders (14750) to their rightful place in the sky dropping bombs. They now follow a (sniffed) spline path as they do on Classic. In support of this, I also created a new override of the `MotionMaster::MoveSplinePath` function to allow a path ID to be passed. I am especially interested in other ideas on how to make this path looped - my solution works but is inelegant.

Since the two flavors of the Bat Riders have very different AI actions, I removed the SAI entirely and re-implemented it in the `npc_batrider` script. The script will detect if it's a summon from Jeklik (BAT_RIDER_MODE_BOSS) or not (BAT_RIDER_MODE_TRASH) and act accordingly.

This is my first AI PR, so I'm looking for feedback on structure, style, and implementation details. I definitely have interest in improving the other fights in ZG as well.

Noteable changes:
- "Boss" Bat Riders are now the appropriate creature, and are limited to 2
- "Boss" Bat Riders follow a spline waypoint path as they do on Classic
- Adjusted bomb throw rate to every 8 seconds to match sniffed data
- "Trash" Bat Riders SAI replaced with script
- Jeklik now reliably returns home and does her green channel effect
- Jeklik will charge the closest player at least 8 yards away during phase 1 (was: random target, any distance)
- `MotionMaster::MoveSplinePath` function added

Special thanks to...
- @heyitsbench for providing the sniff that this PR is based on
- @55Honey for answering questions and pointing me towards Kara as a good starting template
- @AnchyDev for development guidance and ideas
- @Helias for (hopefully) increasing the chances this PR is correct!

## SOURCE:

The changes have been validated through:
- [X] Live research (checked on live servers, e.g Classic WotLK, Retail, etc.)
- [X] Sniffs (remember to share them with the open source community!)
- [X] Video evidence, knowledge databases or other public sources (e.g forums, Wowhead, etc.)
- [ ] The changes promoted by this pull request come partially or entirely from another project (cherry-pick). **Cherry-picks must be committed using the proper --author tag in order to be accepted, thus crediting the original authors, unless otherwise unable to be found**

## Tests Performed:

This PR has been:
- [X] Tested in-game by the author.
- [ ] Tested in-game by other community members/someone else other than the author/has been live on production servers.
- [ ] This pull request requires further testing and may have edge cases to be tested.

## How to Test the Changes:

- [ ] This pull request can be tested by following the reproduction steps provided in the linked issue
- [ ] This pull request requires further testing. Provide steps to test your changes. If it requires any specific setup e.g multiple players please specify it as well.

Jeklik Test:
1.`.go xyz -12261.517578 -1411.231934 130.803345`
2. Clear adds around the boss
4. Engage the boss in honorable combat
5. Observe events firing in phase 1. Charge will only fire if a threat table target is at least 8 yards away.
6. Reduce boss health to <50%
7. Observe transform to humanoid form and normal events firing.
8. Observe two bat riders appearing from the cave and flying their route, throwing bombs. No more than two bat riders appear.
9. `.gm on` then `.gm off` to reset the boss and observe her disappearing before teleporting to the cave and channeling again

Bat Rider Test:
1. Engage any bat rider in a pack on the ground
2. Observe the bat rider performing actions similar to the previous SAI
3. Reduce bat rider to 30% and observe it blow itself up as expected

## Known Issues and TODO List:

None.

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by Nyeriah** [comment]
We don't delete/insert from creature_template, only update the values we are changing

**Comment by kjack9** [comment]
Moving to draft, I have a few issues to work out still it seems.

**Comment by kjack9** [comment]
Some evade edge cases were giving me fits, so I replaced the run back/teleport sequence with a fade out, and then fade back at home sequence. Still better than the previous instant teleport.

Should be good to review again. Looking forward to comments and feedback.

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp [hunk 1]
```diff
@@ -16,19 +16,27 @@
...
 
-enum Says
+enum BossSays
 {
     SAY_AGGRO                   = 0,
     SAY_CALL_RIDERS             = 1,
     SAY_DEATH                   = 2,
     EMOTE_SUMMON_BATS           = 3,
-    EMOTE_GREAT_HEAL            = 4
+    EMOTE_GREAT_HEAL            = 4,
```
Comment (heyitsbench) [inline src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp:34]:
```suggestion
    EMOTE_GREAT_HEAL            = 4
```
Comma added unnecessarily?

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp [hunk 2]
```diff
@@ -97,57 +122,86 @@ Position const SpawnBat[6] =
...
 
-enum Misc
+Position const SpawnBatRider = { -12301.689, -1371.2921, 145.09244 };
+Position const JeklikCaveHomePosition = { -12291.9f, -1380.08f, 144.902f, 2.28638f };
+
+enum PathID
 {
-    PATH_JEKLIK_INTRO = 145170
+    PATH_JEKLIK_INTRO = 145170,
+    PATH_BAT_RIDER_LOOP = 147500
```
Comment (heyitsbench) [inline src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp:131]:
```suggestion
    PATH_JEKLIK_INTRO   = 145170,
    PATH_BAT_RIDER_LOOP = 147500
```
Enums out of line with each other.

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp [hunk 3]
```diff
@@ -174,87 +226,162 @@ struct boss_jeklik : public BossAI
...
                     events.ScheduleEvent(EVENT_GREATER_HEAL, 25s, PHASE_TWO);
                     break;
                 case EVENT_SPAWN_FLYING_BATS:
-                    Talk(SAY_CALL_RIDERS);
-                    if (Unit* target = SelectTarget(SelectTargetMethod::Random, 0))
-                        if (Creature* flyingBat = me->SummonCreature(NPC_FRENZIED_BAT, target->GetPositionX(), target->GetPositionY(), target->GetPositionZ() + 15.0f, 0.0f, TEMPSUMMON_TIMED_DESPAWN_OUT_OF_COMBAT, 15000))
-                            flyingBat->AI()->DoZoneInCombat();
-                    events.ScheduleEvent(EVENT_SPAWN_FLYING_BATS, 10s, 15s, PHASE_TWO);
+                    LOG_DEBUG("scripts.ai", "Jeklik: EVENT_SPAWN_FLYING_BATS");
+                    if (SelectTarget(SelectTargetMethod::Random, 0))
```
Comment (heyitsbench) [inline src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp:364]: Correct me if I'm wrong, but this feels like it's just checking if Jeklik is in combat, in which case I feel like there must be a better way to accomplish it than this.

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp [hunk 4]
```diff
@@ -51,15 +59,22 @@ enum Spells
...
     SPELL_THROW_LIQUID_FIRE     = 23970,
-    SPELL_SUMMON_LIQUID_FIRE    = 23971
+    SPELL_SUMMON_LIQUID_FIRE    = 23971,
+
+    // Bat Rider (Trash)
+    SPELL_DEMO_SHOUT            = 23511,
+    SPELL_BATTLE_COMMAND        = 5115,
+    SPELL_INFECTED_BITE         = 16128,
+    SPELL_PASSIVE_THRASH        = 8876,
+    SPELL_UNSTABLE_CONCOTION    = 24024
```
Comment (heyitsbench) [inline src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp:71]:
```suggestion
    SPELL_UNSTABLE_CONCOCTION    = 24024
```
Non-Blizzlike typo in this spell enum. Will need replacing across the script.

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp (hunk 3 above)
Comment (kjack9) [reply src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp:364]: Close - engaging the boss puts the entire zone in combat. However, I believe `SelectTarget` only pulls from the list of players on the threat table.

I did just follow the lead of the original author here, but I'm glad to pivot if checking just for InCombat is better.

Comment (kjack9) [reply src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp:364]: Would this be clearer? I'll need to test it, of course.

`if (me->GetThreatMgr().GetThreatListSize() > 0)`

Comment (heyitsbench) [reply src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp:364]: I'm just asking from a performant standpoint. The previous implementation actually used the selected target as a part of the script, whereas this one just seems, like I said, to check for valid combat targets.

> Would this be clearer?

Might be able to knock off the `> 0` and the evaluation will work the same, but yes, it does seem clearer to me at least.

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp [hunk 5]
```diff
@@ -78,7 +93,17 @@ enum Events
...
+
+    // Bat Riders (Boss)
+    EVENT_BAT_RIDER_LOOP,
+    EVENT_BAT_RIDER_THROW_BOMB,
+
+    // Bat Riders (Trash)
+    EVENT_DEMO_SHOUT,
+    EVENT_BATTLE_COMMAND,
+    EVENT_INFECTED_BITE,
+    EVENT_UNSTABLE_CONCOTION
```
Comment (heyitsbench) [inline src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp:106]:
```suggestion
    EVENT_UNSTABLE_CONCOCTION
```
Don't forget here, as well as L523, L526, L546, and L547.

**Comment by kjack9** [comment]
Thank you for all the helpful comments, I believe I have addressed them all. Let me know what else needs to be updated.

**Review by heyitsbench — APPROVED** [review]

**Comment by kjack9** [comment]
I reduced the amount of logging and standardized it to what I see in other parts of AC. I also implemented a safety check to ensure that the bat riders counter was getting reset with the boss.

I've tested this with my personal group now and it seems to be working well. Is there anything else I can address for this encounter?

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp [hunk 6]
```diff
@@ -265,45 +387,155 @@ struct boss_jeklik : public BossAI
...
-private:
-    TaskScheduler _scheduler;
+            // if health goes below 30%, cast Unstable Concotion
+            if (me->HealthBelowPct(30))
+            {
+                events.ScheduleEvent(EVENT_UNSTABLE_CONCOCTION, 0);
+            }
+
+            // event handling
+            switch (events.ExecuteEvent())
```
Comment (elthehablo) [inline src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp:515]: why not create a private TaskScheduler and make the bat rider abilities fire with the taskscheduler as well?

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp [hunk 7]
```diff
@@ -265,45 +387,155 @@ struct boss_jeklik : public BossAI
...
+                    DoCast(me, SPELL_BATTLE_COMMAND);
+                    events.ScheduleEvent(EVENT_BATTLE_COMMAND, 25s);
+                    break;
+                case EVENT_INFECTED_BITE:
+                    DoCastVictim(SPELL_INFECTED_BITE);
+                    events.ScheduleEvent(EVENT_INFECTED_BITE, 8s);
+                    break;
+                case EVENT_UNSTABLE_CONCOCTION:
+                    Talk(EMOTE_LOW_HEALTH);
+                    DoCast(me, SPELL_UNSTABLE_CONCOCTION);
```
Comment (elthehablo) [inline src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp:530]:
```suggestion
                    DoCastSelf(SPELL_UNSTABLE_CONCOCTION);
```

and so on for the others as well

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp [hunk 8]
```diff
@@ -265,45 +387,155 @@ struct boss_jeklik : public BossAI
...
+        {
+            case BAT_RIDER_MODE_BOSS:
+                events.Reset();
+                me->GetMotionMaster()->Clear();
+                break;
+            case BAT_RIDER_MODE_TRASH:
+                events.Reset();
+
+                // apply the Thrash (8876) spell to the bat rider (passive ability)
+                me->CastSpell(me, SPELL_PASSIVE_THRASH);
```
Comment (elthehablo) [inline src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp:445]: add this as aura in the db. `creature_template_addon`

**Review by elthehablo — CHANGES_REQUESTED** [review]
Nice work overall!

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp [hunk 9]
```diff
@@ -174,32 +232,72 @@ struct boss_jeklik : public BossAI
...
 
         while (uint32 eventId = events.ExecuteEvent())
         {
             switch (eventId)
             {
-                // Phase one
+                // PHASE_ONE
                 case EVENT_CHARGE_JEKLIK:
-                    if (Unit* target = SelectTarget(SelectTargetMethod::Random, 0))
+                    // charge the nearest player that is at least 8 yards away (charge min distance)
```
[the remark this answers is not in the document]
Comment (kjack9) [reply src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp:290]: I'll be happy to change this, but I'd like to understand why this case needs a scheduler and the rest is okay being event-based? My understanding from asking on Discord is that either approach is acceptable, and since the rest of the AI is done using events I stuck to that.

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp (hunk 8 above)
Comment (kjack9) [reply src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp:445]: Are you okay with the flying boss version (incorrectly) having this aura too then? Would it make a difference if I moved this to initialization and only cast it once?

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (elthehablo) [reply src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp:290]: Yes. But I don't think seeing both in 1 script makes a lot of sense. So it's best to pick one. I would go with the scheduler, personally, as you're essentially revamping most of the script. Taskscheduler is the new standard for boss scripts

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp (hunk 8 above)
Comment (elthehablo) [reply src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp:445]: if the adds are pre-spawned, you can also potentially add it to `creature_addon` so it only applies to the creatures that are pre-spawned. but nah this is fine then imo

**Comment by kjack9** [comment]
Alright, this is ready again. Completely removed the event system and replaced with TaskScheduler. Everything is working well in testing, I can't find a way to break it. Let me know if anything else needs to change.

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (kjack9) [reply src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp]: Done, events are no more.

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp (hunk 7 above)
Comment (kjack9) [reply src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp]: Done.

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp (hunk 6 above)
Comment (kjack9) [reply src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp]: Done.

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp [hunk 10]
```diff
@@ -97,213 +92,409 @@ Position const SpawnBat[6] =
...
-            events.ScheduleEvent(EVENT_SHADOW_WORD_PAIN, 10s, 15s, PHASE_TWO);
-            events.ScheduleEvent(EVENT_PSYCHIC_SCREAM, 25s, 35s, PHASE_TWO);
-            events.ScheduleEvent(EVENT_MIND_FLAY, 10s, 30s, PHASE_TWO);
-            events.ScheduleEvent(EVENT_GREATER_HEAL, 25s, PHASE_TWO);
-            events.ScheduleEvent(EVENT_SPAWN_FLYING_BATS, 10s, PHASE_TWO);
+            scheduler.CancelGroup(PHASE_ONE);
 
-            return;
-        }
+            // Curse of Blood
```
Comment (heyitsbench) [inline src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp:266]: Not sure on precedent, but these comments feel redundant given the enumeration is being utilized.

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp [hunk 11]
```diff
@@ -97,213 +92,409 @@ Position const SpawnBat[6] =
...
-        if (!UpdateVictim())
-            return;
+        // enter evade mode
+        BossAI::EnterEvadeMode(why);
 
-        events.Update(diff);
+        if (why != EvadeReason::EVADE_REASON_NO_PATH)
+        {
+            // make invisible to hide wonky-looking movement
+            me->SetVisible(false);
```
Comment (heyitsbench) [inline src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp:327]: This feels...hacky, though I don't know of a better way to correct it.

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp [hunk 12]
```diff
@@ -97,213 +92,409 @@ Position const SpawnBat[6] =
...
+            me->SetCanFly(true);
+            me->SetDisableGravity(true);
+            me->GetMotionMaster()->MoveSplinePath(PATH_BATRIDER_LOOP);
+        }
+        // otherwise, trash mode
+        else
+        {
+            LOG_DEBUG("scripts.ai", "npc_batrider::constructor: BATRIDER_MODE_TRASH");
+            me->SetReactState(REACT_DEFENSIVE);
+            _mode = BATRIDER_MODE_TRASH;
```
Comment (heyitsbench) [inline src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp:394]: Personal grievance: maybe move this `_mode =` directly under the log as it is in the boss summon block for consistency.

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp [hunk 13]
```diff
@@ -97,213 +92,409 @@ Position const SpawnBat[6] =
...
+            // throw bomb
+            _scheduler.Schedule(2s, [this](TaskContext context)
+            {
+                DoCastRandomTarget(SPELL_BATRIDER_THROW_LIQUID_FIRE);
+                context.Repeat(8s);
+            });
+        }
+        else if (_mode == BATRIDER_MODE_TRASH)
+        {
+            // demo shout
```
Comment (heyitsbench) [inline src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp:436]: The styling of this scheduler block is quite different from Jeklik's scheduling. This block has spacing between each spell, plus lowercase spell name comments. I personally prefer Jeklik's styling.

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp (hunk 11 above)
Comment (kjack9) [reply src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp:327]: Couldn't agree more but after a few hours of messing with movement I can't get her to run back to a ground home position before teleporting herself up to the cave. There is no path to get up there, so a smooth transition is impossible.

To add to the pain, if she evades while gliding down from her cave, she tends to get caught halfway through the spline path and I've found it nearly impossible to cancel it cleanly. It MAY be related to [#17595](https://github.com/azerothcore/azerothcore-wotlk/issues/17595), but regardless I can't make it look right.

This solution isn't too bad from the player's PoV - she does a reasonable-length fade-out, and then several seconds later appears in her home spot casting the right thing. It is the best of the options I was able to figure out how to do.

TL;DR - I tried. I failed. This doesn't suck.

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp (hunk 10 above)
Comment (kjack9) [reply src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp:266]: I guess this is my replacement for the clean view of the EventMap where I could easily see all the abilities together. At least this way all of them are in green text and I can quickly identify them. I find it visually difficult to track down the one enum value (when multiple are used) to get to the spell I care about.

If it's a show-stopper, I can remove them. IMO ACore could use some more comments for newbs like me trying to learn how this thing works, but I understand that's not the culture here and I'm trying to find the right balance. See: very few logs in this AI.

Let me know what you'd like me to do.

Comment (kjack9) [reply src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp:266]: This is what I'm talking about with EventMap - super easy to see the timing settings, and I can go right to the spell I care about. TaskScheduler makes that difficult and at least for me, the comments create structure that's not inherent in the TaskScheduler way.

image

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp (hunk 13 above)
Comment (kjack9) [reply src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp]: Fixed.

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp (hunk 12 above)
Comment (kjack9) [reply src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp:393]: Done.

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp (hunk 11 above)
Comment (kjack9) [reply src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp:327]: I'm going to resolve this one for now, please re-open if you feel it merits further discussion.

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp (hunk 10 above)
Comment (elthehablo) [reply src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp:266]: Yeah agree with Bench here. Imo they shouldn't be needed when the spells already use the enum

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp [hunk 14]
```diff
@@ -97,213 +92,406 @@ Position const SpawnBat[6] =
...
+    void DamageTaken(Unit* /*attacker*/, uint32& damage, DamageEffectType, SpellSchoolMask) override
     {
-        if (!UpdateVictim())
-            return;
-
-        _scheduler.Update(diff);
+        if (_mode == BATRIDER_MODE_TRASH)
+        {
+            // if health goes below 30%, cast unstable concoction
+            if (me->HealthBelowPctDamaged(30, damage))
```
Comment (elthehablo) [inline src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp:460]: ScheduleHealthCheckEvent to remain consistent

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp (hunk 10 above)
Comment (kjack9) [reply src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp:266]: I understand why namespaces got objected to since it was a new coding practice, but I must admit that I don't understand the objection here. As a newb reading an AI for the first time, it would have made a huge difference to me if someone commented and explain WTF they were doing instead of just assuming I understood all their code.

Maybe I'm just not cut out to redo these if all the things I feel like improve the code aren't acceptable. Very frustrated.

I will remove them if that's what's required to get this accepted.

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp (hunk 14 above)
Comment (kjack9) [reply src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp:460]: Not available in `CreatureAI`. Do you want me to convert to BossAI?

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp (hunk 10 above)
Comment (heyitsbench) [reply src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp:266]: Quite frankly, I feel it's just a personal preference thing and definitely don't think it should be a barrier to merge. I mean, it is just a comment after all, we've all seen worse comments in this codebase. At the end of the day, I certainly don't mind 'em being there.

Comment (kjack9) [reply src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp]: I pushed a commit that removes any comments that might be considered obvious. Let me know if that is closer to the level of commenting I should be aiming for.

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp [hunk 15]
```diff
@@ -97,213 +92,371 @@ Position const SpawnBat[6] =
...
+                DoCastSelf(SPELL_GREATER_HEAL);
+                context.Repeat(25s);
+            }).Schedule(10s, PHASE_TWO, [this](TaskContext context)
+            {
+                if (me->GetThreatMgr().GetThreatListSize())
+                {
+                    // summon up to 2 bat riders
+                    if (batRidersCount < 2)
+                    {
+                        LOG_DEBUG("scripts.ai", "boss_jeklik::UpdateAI:: Spawn Flying Bats (Summoning {} of 2)", batRidersCount + 1);
```
Comment (elthehablo) [inline src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp:284]:
```suggestion

```

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp [hunk 16]
```diff
@@ -97,213 +92,371 @@ Position const SpawnBat[6] =
...
+            me->GetEntry() == NPC_BATRIDER &&
+            me->IsSummon() &&
+            me->ToTempSummon() &&
+            me->ToTempSummon()->GetSummoner() &&
+            me->ToTempSummon()->GetSummoner()->GetEntry() == NPC_PRIESTESS_JEKLIK
+        )
         {
-            return !me->HasUnitState(UNIT_STATE_CASTING);
-        });
+            LOG_DEBUG("scripts.ai", "npc_batrider::constructor: BATRIDER_MODE_BOSS");
```
Comment (elthehablo) [inline src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp:351]:
```suggestion

```

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp [hunk 17]
```diff
@@ -97,213 +92,371 @@ Position const SpawnBat[6] =
...
+            me->SetReactState(REACT_PASSIVE);
+
+            me->SetSpeed(MOVE_WALK, 5.0f, true);
+
+            me->SetCanFly(true);
+            me->GetMotionMaster()->MoveSplinePath(PATH_BATRIDER_LOOP);
+        }
+        else
+        {
+            LOG_DEBUG("scripts.ai", "npc_batrider::constructor: BATRIDER_MODE_TRASH");
```
Comment (elthehablo) [inline src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp:367]:
```suggestion

```

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp [hunk 18]
```diff
@@ -97,213 +92,371 @@ Position const SpawnBat[6] =
...
 
-private:
-    TaskScheduler _scheduler;
+    void UpdateAI(uint32 /*diff*/) override
+    {
+        if (_mode == BATRIDER_MODE_BOSS)
+        {
+            if (!me->isMoving())
+            {
+                LOG_DEBUG("scripts.ai", "npc_batrider::UpdateAI: not moving, running loop");
```
Comment (elthehablo) [inline src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp:443]:
```suggestion

```

**Review by elthehablo — CHANGES_REQUESTED** [review]
please remove all debug lines. this is just something we do not want to have in general scripting in AzerothCore. I know you disagree with this, but it is just a precedent we don't want to set. Otherwise the fight seems to work well when I tested it so should be good to merge

**Comment by kjack9** [comment]
> please remove all debug lines. this is just something we do not want to have in general scripting in AzerothCore. I know you disagree with this, but it is just a precedent we don't want to set. Otherwise the fight seems to work well when I tested it so should be good to merge

I know when to give up. All logging and most comments have been removed as of the latest commit. Re-tested just to be sure and everything seems to be working fine.

**Comment by elthehablo** [comment]
yeah thanks for that, man :)

**Review by elthehablo — APPROVED** [review]

**Comment by elthehablo** [comment]
good work on this. keep it up
